Spec for MSC3824: OAuth 2.0 aware clients (#2272)
matrix-org/matrix-spec: content/client-server-api/_index.md

user management by [application services](application-service-api/#server-admin-style-permissions).
{{% /boxes/note %}}

{{% boxes/note %}}
{{% added-in v="1.18" %}}
A compatibility layer is available to ease the transition from the legacy API to
the OAuth 2.0 API for clients that only support the legacy API called [OAuth 2.0

> aware clients](#oauth-20-aware-clients).

Reviewer: ```suggestion
A compatibility feature, called [OAuth 2.0 aware clients](#oauth-20-aware-clients), is available to ease
the transition to the OAuth 2.0 API for clients that only support the legacy API.
```

is complete, the client will need to submit a `/login` request matching
`m.login.token`.

{{% added-in v="1.18" %}} [OAuth 2.0 aware clients](/client-server-api/#oauth-20-aware-clients)
MUST only offer the `m.login.sso` flow to the user when `oauth_aware_preferred`
is set to `true` and MUST add the `action=login` parameter to the SSO redirect

> endpoint.

Reviewer: I'm not really sure we need to say this here as well as everywhere else.

```suggestion
```

{{% http-api spec="client-server" api="account_deactivation" %}}

#### OAuth 2.0 aware clients

{{% added-in v="1.18" %}}

> This is a compatibility layer that allows clients that only support the legacy

Reviewer 2: I don't think that "layer" is appropriate to describe this as I don't think it is a layer in a Computer Science sense.

Ideas for possible alternatives: feature; mode;

Also, maybe we could start with a shorter sentence to give the context.

```suggestion
This is a compatibility feature to aide clients in the transition to the OAuth 2.0 API. It allows clients that only support the legacy
```

Author: Applied in e5ad585bfdd9b5125de17c129710a5c9adf4248f.

API to make some less-invasive changes to improve the user experience when
talking to a homeserver that is using the OAuth 2.0 API without actually having
to implement the full OAuth 2.0 API.

##### Client behaviour

For a client to be considered fully OAuth 2.0 aware it MUST:

* Support the [`m.login.sso` authentication flow](#client-login-via-sso).
* Where a `oauth_aware_preferred` value of `true` is present on an `m.login.sso`
  flow then *only* offer that auth flow to the user.

> * Append `action=login` and `action=register` parameters to the [SSO redirect

Reviewer: ```suggestion
* Append `action=login` or `action=register` parameters to the [SSO redirect
```

Author: Applied in e5ad585bfdd9b5125de17c129710a5c9adf4248f.
